utils/promoteMavenArtifacts.py

```diff
@@ -0,0 +1,197 @@
+
+    print(payload)
+
+    url = f"{ URL }/api/search/aql"
+
+    response = requests.post(url,
+                             auth=requests.auth.HTTPBasicAuth(
+                                 USERNAME,
+                                 PASSWORD),
+                             data=payload)
+
+    j = response.json()
+
+    for result in j["results"]:
+        path = "/" + result["path"]
+        if path not in directories:
+            directories.append(path)
+
+    return directories
+
+
+def is_directory_exist(repository, directory):
+    """
+        is_directory_exist query a maven repository
+        to see if a directory already exist
+    """
+    url = URL + '/api/storage/' + repository + directory
+
+    response = requests.get(url,
+                            auth=requests.auth.HTTPBasicAuth(
+                                USERNAME,
+                                PASSWORD))
+
+    if response.status_code != 200:
+        return False
+
+    return response.status_code == 200
+
+
+def is_alive():
+    """
+        is_alive test if a maven repository is reachable based on a ping test
+    """
+    url = URL + '/api/system/ping'
+
+    response = requests.get(url,
+                            auth=requests.auth.HTTPBasicAuth(
+                                USERNAME,
+                                PASSWORD))
+
+    return response.status_code == 200
+
+
+def is_required_parameters():
+    """
+        is_required_parameters check if all settings are provided
+    """
+    result = True
+
+    if not URL:
+        print("Missing environment variable MAVEN_REPOSITORY_URL")
+        result = False
+    if not USERNAME:
+        print("Missing environment variable MAVEN_REPOSITORY_USERNAME")
+        result = False
+    if not PASSWORD:
+        print("Missing environment variable MAVEN_REPOSITORY_USERNAME")
+        result = False
+    if not srcRepoKey:
+        print("Missing environment variable MAVEN_REPOSITORY_SOURCE_NAME")
+        result = False
+    if not targetRepoKey:
+        print("Missing environment variable MAVEN_REPOSITORY_TARGET_NAME")
+        result = False
+    if not VERSION:
+        print("Missing environment variable JENKINS_VERSION")
+        result = False
+
+    return result
+
+
+if __name__ == "__main__":
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument("version",
+                        help="specify version that need to be promoted")
+    parser.add_argument("--dry_run",
+                        help="Don't copy items",
+                        action="store_true")
+
+    args = parser.parse_args()
+
+    VERSION = args.version
+    dryrun = int(args.dry_run)
+
+    if not is_required_parameters():
+        sys.exit(1)
+
+    if not is_alive():
+        print(f"{ URL } not reachable")
+        sys.exit(2)
+
+    print(f"Artifactory version: {get_api_version()}\n")
+
+    directories = get_directories(srcRepoKey, "/org/jenkins-ci/main", VERSION)
```

Comment:
The approach of copying individual artifacts seems unnecessarily brittle as it makes possibly invalid assumptions about groupIds. It's also not atomic on the client side.

Reply:
Are you suggesting to do a full copy of the repository so instead of copying 'jenkins-staging/org/jenkins-ci/main/jenkins-war/2.232' to 'releases/org/jenkins-ci/main/jenkins-war/2.232', we would do copy 'jenkins-staging' to 'releases'?

Reply:
Yes. That's how I've been doing security staging for years.

Granted, I have even less of an idea how what will be included, but it works well enough (except having to compute the `maven-metadata.xml` afterwards).

Note that this only makes sense if you don't copy the same thing over and over, so if you want to copy instead of move, remember to empty out the staging repo at the end of the process.

Reply:
> The approach of copying individual artifacts seems unnecessarily brittle as it makes possibly invalid assumptions about groupIds. It's also not atomic on the client-side.

While it's not an atomic operation, it can be run as many times as we want since it only copies files that don't exist on the remote repository (cfr, output in this PR description), and we could also hardcode groupIds instead of trying to guess them. But I agree this is  the weakest part that I am not really happy with
The risk of trying to guess the groupsIDs is to promote unwanted items, like remoting, but this is mitigated if the staging repository uses only contains core releases
The risk of hardcoding groupsIds is to forget some groupIDs in the future.

Using the approach where we copy the full repository, means that we have no control of what we copy so we could override valid items on the production repository. So must ensure that the source repository only contains artifacts for the release that we want to release and we use move instead of copy, while in both cases they always override destination items, move also take care of removing source items.

Reply:
> Using the approach where we copy the full repository, means that we have no control of what we copy so we could override valid items on the production repository. So must ensure that the source repository only contains artifacts for the release that we want to release and we use move instead of copy, while in both cases they always override destination items, move also take care of removing source items.

Agree, all of the available options aren't great.

One potential solution is to not grant Delete permission, which means we don't allow overriding either, which should result in (deserved) failures. I'm unsure how this will handle the replacement of `maven-metadata.xml` though, would need testing. In fact I would recommend we do not grant the user involved delete permission either way.

Reply:
I couldn't find a good solution that I am happy with so I decided to add each of them and let the release maintainer select the one he prefers.
`promoteMavenArtifacts.py` can be used in three different ways

1. Copy/Move a full repository content to target and then recursively update metadata.xml
Downside: it will override destination items if they already exist, theorically this can be mitigated by not having 'delete' permission on the remote repository.
This assumption still need to be validated especially for metadata.xml

2. Copy/Move every item for a specific  version under a specific path
 Thre 
Downside: There is a risk to copy/move unwanted items if it matches the version that needs to be copied.
 
3. Copy/Move every specific item for groupIDs provided by parameters
Downside: We have to manually maintain the list of item that needs to be promoted

### Every item for a specific version

```
      ./utils/promoteMavenArtifacts.py item --url https://repo.jenkins-ci.org --username username --password xxx  --source releases  --destination sandbox --search '/org/jenkins-ci/main' 2.224
        Following directories will be promoted:
        ['/org/jenkins-ci/main/cli/2.224', '/org/jenkins-ci/main/jenkins-bom/2.224', '/org/jenkins-ci/main/jenkins-core/2.224', '/org/jenkins-ci/main/jenkins-parent/2.224', '/org/jenkins-ci/main/jenkins-war/2.224']


        Artifactory version: 6.18.1

        [1/5] - /org/jenkins-ci/main/cli/2.224

        Already exist on sandbox

        [2/5] - /org/jenkins-ci/main/jenkins-bom/2.224

        Already exist on sandbox

        [3/5] - /org/jenkins-ci/main/jenkins-core/2.224

        Already exist on sandbox

        [4/5] - /org/jenkins-ci/main/jenkins-parent/2.224

        Already exist on sandbox

        [5/5] - /org/jenkins-ci/main/jenkins-war/2.224

        Planning to copy '/org/jenkins-ci/main/jenkins-war/2.224'
                  from releases to olblak-sandbox


        https://repo.jenkins-ci.org/api/copy/releases/org/jenkins-ci/main/jenkins-war/2.224?to=/sandbox//org/jenkins-ci/main/jenkins-war&dry=0&suppressLayout=0 0&failFAst=1
        {'messages': [{'level': 'INFO', 'message': 'copying releases:org/jenkins-ci/main/jenkins-war/2.224 to sandbox:org/jenkins-ci/main/jenkins-war completed successfully, 8 artifacts and 1 folders were copied'}]}
        copying releases:org/jenkins-ci/main/jenkins-war/2.224 to sandbox:org/jenkins-ci/main/jenkins-war completed successfully, 8 artifacts and 1 folders were copied
```

### Full repository content
```
 ./utils/promoteMavenArtifacts.py repository --url https://repo.jenkins-ci.org --username username --password password  --source source  --destination sandbox
Artifactory version: 6.18.1

Planning to copy source to sandbox


https://repo.jenkins-ci.org/api/copy/source/?to=/sandbox//&dry=0&suppressLayout=0 0&failFAst=1
{'messages': [{'level': 'INFO', 'message': 'copying source: to sandbox: completed successfully, 213 artifacts and 31 folders were copied'}]}
copying source: to sandbox: completed successfully, 213 artifacts and 31 folders were copied

Every metadata.xml under sandbox/ were successfully updated

```

### Specified item for a specific version

```
     $: ./utils/promoteMavenArtifacts.py item --url https://repo.jenkins-ci.org --username username --password xxx  --source releases  --destination sandbox --groupID '/org/jenkins-ci/main/cli' 2.224

        Artifactory version: 6.18.1

        [1/1] - /org/jenkins-ci/main/cli/2.224

        Already exist on sandbox


As suggested in the security profile, different profiles can have different behaviors.

Reply:
for security releases, I propose to let @daniel-beck manage source repository and move everything as defined in profile.d/security 
`PROMOTE_STAGING_MAVEN_ARTIFACTS_ARGS="repository --mode move --source $MAVEN_REPOSITORY_NAME --destination $MAVEN_REPOSITORY_PRODUCTION_NAME --url $MAVEN_REPOSITORY_URL --username $MAVEN_REPOSITORY_USERNAME --password **$MAVEN_REPOSITORY_PASSWORD"`

Reply:
In order to move a full repository, we need 'delete' permission 

https://repo.jenkins-ci.org/api/move/sandbox/?to=/sandbox-2//&dry=0&suppressLayout=0 0&failFAst=1
{'messages': [{'level': 'ERROR', 'message': "User doesn't have permissions to override 'sandbox-2:'. Needs delete permissions."}]}
User doesn't have permissions to override 'sandbox-2:'. Needs delete permissions.